### PR TITLE
Improve JS performance by storing length before comparing to it in loops

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Since https://github.com/rust-lang/rust/pull/79052 is quite complicated to review, I suggested to split into smaller parts. This first part is mostly about saving the array length into a variable (I tried to not change anything else as much as possible :smiley: ).

r? @jyn514

### CODE REVIEW

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -903,12 +901,10 @@ function defocusSearchBar() {
...
-                        var length = obj[GENERICS_DATA].length;
-                        for (x = 0; x < length; ++x) {
-                            if (obj[GENERICS_DATA][x] === val.name) {
-                                return true;
-                            }
-                        }
+                        return onEach(obj[GENERICS_DATA],
+                            function(name) {
+                                return name === val.name;
+                            });
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:907]:
```suggestion
                        return obj[GENERICS_DATA].any(
                            function(name) {
                                return name === val.name;
                            });
```

**Review by jyn514 — COMMENTED** [review]
I think the new code isn't any easier to read, and it creates a lot of churn.

File: src/librustdoc/html/static/main.js (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:907]: You mean `some` I guess? ;)

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:907]: And some is indeed faster! I tested on jsbench again the two codes:

```js
function onEach(arr, func) {
	for (var i = 0, len = arr.length; i < len; ++i) {
		if (func(arr[i]) === true) {
			return true;
		}
	}
	return false;
}

var arr = ["a", "bcdef", "ghi", "ksdf", "jsqdfkqsdfq", "qdjkfqdq", "dfkqjdnfq", "qdjkfqdjkfq", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd",
"qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd",
"qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd",
"qfqdjfnqd", "dada"];
var found = onEach(arr, function(el) { return el === "dada"});
```

and:

```js
var arr = ["a", "bcdef", "ghi", "ksdf", "jsqdfkqsdfq", "qdjkfqdq", "dfkqjdnfq", "qdjkfqdjkfq", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd",
"qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd",
"qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd", "qfqdjfnqd",
"qfqdjfnqd", "dada"];
var found = arr.some(function(el) { return el === "dada"});
```

The first one is ~8.47 slower. So let's go for `some` and thanks a lot for the suggestion!

File: src/librustdoc/html/static/main.js [hunk 2]
```diff
@@ -830,14 +827,13 @@ function defocusSearchBar() {
...
-                            for (var x = 0; x < elength; ++x) {
-                                var tmp_lev = levenshtein(getObjectFromId(elems[x]).name,
-                                                          firstGeneric);
+                        var x, lev, elength, firstGeneric, tmp_lev;
+                        for (var y = 0, vlength = val.generics.length; y < vlength; ++y) {
+                            lev = { pos: -1, lev: MAX_LEV_DISTANCE + 1};
+                            elength = elems.length;
+                            firstGeneric = getObjectFromId(val.generics[y]).name;
+                            for (x = 0, elength = elems.length; x < elength; ++x) {
+                                tmp_lev = levenshtein(getObjectFromId(elems[x]).name, firstGeneric);
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:836]: None, sorry, this change shouldn't have been part of this PR. Removing it.

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/librustdoc/html/static/main.js [hunk 3]
```diff
@@ -835,9 +832,9 @@ function defocusSearchBar() {
                             var lev = { pos: -1, lev: MAX_LEV_DISTANCE + 1};
                             var elength = elems.length;
                             var firstGeneric = getObjectFromId(val.generics[y]).name;
-                            for (var x = 0; x < elength; ++x) {
+                            for (x = 0, elength = elems.length; x < elength; ++x) {
```
Comment (Nemo157) [inline src/librustdoc/html/static/main.js:835]: I don't see any declaration for `x`?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:835]: Good catch!

**Comment by GuillaumeGomez** [comment]
@Nemo157 Updated!

File: src/librustdoc/html/static/main.js [hunk 4]
```diff
@@ -1710,7 +1706,7 @@ function defocusSearchBar() {
                     if (smallest === null) {
                         break;
                     }
-                    for (x = 0; x < arrays.length && ret.length < MAX_RESULTS; ++x) {
+                    for (x = 0; x < arrays_len && ret.length < MAX_RESULTS; ++x) {
```
Comment (Nemo157) [inline src/librustdoc/html/static/main.js:1709]: This is using the non-lexical variable from line 1700, right? JS is fun 😭.

**Review by Nemo157 — APPROVED** [review]
The rest of the changes all look correct 👍

File: src/librustdoc/html/static/main.js (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1709]: Yes. It's a syntax nightmare. ^^'

**Comment by GuillaumeGomez** [comment]
Thanks @Nemo157 !

@bors: r=Nemo157,jyn514 rollup

**Pull request merged**